src/core/ext/filters/client_channel/lb_policy/grpclb/client_load_reporting_filter.cc

```diff
@@ -33,118 +34,48 @@
-                                  const grpc_call_final_info* /*final_info*/,
-                                  grpc_closure* /*ignored*/) {
-  call_data* calld = static_cast<call_data*>(elem->call_data);
-  if (calld->client_stats != nullptr) {
-    // Record call finished, optionally setting client_failed_to_send and
-    // received.
-    calld->client_stats->AddCallFinished(
-        !calld->send_initial_metadata_succeeded /* client_failed_to_send */,
-        calld->recv_initial_metadata_succeeded /* known_received */);
+  // Handle client initial metadata.
+  // Grab client stats object from metadata.
+  auto client_stats_md = call_args.client_initial_metadata->Take(
+      grpc_core::GrpcLbClientStatsMetadata());
+  if (client_stats_md.has_value()) {
+    client_stats.reset(*client_stats_md);
   }
-  calld->~call_data();
-}
 
-static void clr_start_transport_stream_op_batch(
-    grpc_call_element* elem, grpc_transport_stream_op_batch* batch) {
-  call_data* calld = static_cast<call_data*>(elem->call_data);
-  // Handle send_initial_metadata.
-  if (batch->send_initial_metadata) {
-    // Grab client stats object from metadata.
-    auto client_stats_md =
-        batch->payload->send_initial_metadata.send_initial_metadata->Take(
-            grpc_core::GrpcLbClientStatsMetadata());
-    if (client_stats_md.has_value()) {
-      grpc_core::GrpcLbClientStats* client_stats = *client_stats_md;
-      if (client_stats != nullptr) {
-        calld->client_stats.reset(client_stats);
-        // Intercept completion.
-        calld->original_on_complete_for_send = batch->on_complete;
-        GRPC_CLOSURE_INIT(&calld->on_complete_for_send, on_complete_for_send,
-                          calld, grpc_schedule_on_exec_ctx);
-        batch->on_complete = &calld->on_complete_for_send;
-      }
-    }
-  }
-  // Intercept completion of recv_initial_metadata.
-  if (batch->recv_initial_metadata) {
-    calld->original_recv_initial_metadata_ready =
-        batch->payload->recv_initial_metadata.recv_initial_metadata_ready;
-    GRPC_CLOSURE_INIT(&calld->recv_initial_metadata_ready,
-                      recv_initial_metadata_ready, calld,
-                      grpc_schedule_on_exec_ctx);
-    batch->payload->recv_initial_metadata.recv_initial_metadata_ready =
-        &calld->recv_initial_metadata_ready;
-  }
-  // Chain to next filter.
-  grpc_call_next_op(elem, batch);
+  auto* server_initial_metadata = call_args.server_initial_metadata;
+
+  return Seq(next_promise_factory(std::move(call_args)),
+             [server_initial_metadata, client_stats = std::move(client_stats)](
+                 ServerMetadataHandle trailing_metadata) {
+               if (client_stats != nullptr) {
+                 client_stats->AddCallFinished(
+                     trailing_metadata->get(GrpcStreamNetworkState()) ==
+                         GrpcStreamNetworkState::kNotSentOnWire,
+                     NowOrNever(server_initial_metadata->Wait()).has_value());
```

Comment:
Do we have a guarantee that `server_initial_metadata` will still exist at this point?

Also, I assume there's no problem with two different callers calling `Wait()` on the same latch?

Reply:
The latch is always on the arena, so yeah... all good for that.
And no problem with multiple callers calling `Wait()` on the same latch.